src/shogun/machine/gp/FITCInferenceMethod.cpp

```diff
@@ -317,74 +306,102 @@ void CFITCInferenceMethod::update_deriv()
 	eigen_w=eigen_B*eigen_al;
 
 	// create shogun and eigen representation of W
-	m_W=SGMatrix<float64_t>(m_chol_utr.num_cols, m_dg.vlen);
-	Map<MatrixXd> eigen_W(m_W.matrix, m_W.num_rows, m_W.num_cols);
+	m_Rvdd=SGMatrix<float64_t>(m_chol_utr.num_cols, m_t.vlen);
+	Map<MatrixXd> eigen_W(m_Rvdd.matrix, m_Rvdd.num_rows, m_Rvdd.num_cols);
 
 	// compute W=Lu'\(V./repmat(g_sn2',nu,1))
+	//W = Lu'\(V./repmat(g_sn2',nu,1));
 	eigen_W=eigen_Lu.triangularView<Upper>().adjoint().solve(V*VectorXd::Ones(
-		m_dg.vlen).cwiseQuotient(eigen_dg).asDiagonal());
+		m_t.vlen).cwiseProduct(eigen_t).asDiagonal());
 }
 
-SGVector<float64_t> CFITCInferenceMethod::get_derivative_wrt_inference_method(
-		const TParameter* param)
-{
-	REQUIRE(!strcmp(param->m_name, "scale"), "Can't compute derivative of "
-			"the nagative log marginal likelihood wrt %s.%s parameter\n",
-			get_name(), param->m_name)
-
-	// create eigen representation of dg, al, B, W, w
-	Map<VectorXd> eigen_dg(m_dg.vector, m_dg.vlen);
-	Map<VectorXd> eigen_al(m_al.vector, m_al.vlen);
-	Map<VectorXd> eigen_w(m_w.vector, m_w.vlen);
-	Map<MatrixXd> eigen_B(m_B.matrix, m_B.num_rows, m_B.num_cols);
-	Map<MatrixXd> eigen_W(m_W.matrix, m_W.num_rows, m_W.num_cols);
-
-	// clone kernel matrices
-	SGVector<float64_t> deriv_trtr=m_ktrtr.get_diagonal_vector();
-	SGMatrix<float64_t> deriv_uu=m_kuu.clone();
-	SGMatrix<float64_t> deriv_tru=m_ktru.clone();
 
-	// create eigen representation of kernel matrices
-	Map<VectorXd> ddiagKi(deriv_trtr.vector, deriv_trtr.vlen);
-	Map<MatrixXd> dKuui(deriv_uu.matrix, deriv_uu.num_rows, deriv_uu.num_cols);
-	Map<MatrixXd> dKui(deriv_tru.matrix, deriv_tru.num_rows, deriv_tru.num_cols);
-
-	// compute derivatives wrt scale for each kernel matrix
-	ddiagKi*=m_scale*2.0;
-	dKuui*=m_scale*2.0;
-	dKui*=m_scale*2.0;
+SGVector<float64_t> CFITCInferenceMethod::get_posterior_mean()
+{
+	if (parameter_hash_changed())
+		update();
 
-	// compute R=2*dKui-dKuui*B
-	MatrixXd R=2*dKui-dKuui*eigen_B;
+	m_mu=SGVector<float64_t>(m_al.vlen);
+	Map<VectorXd> eigen_mu(m_mu.vector, m_mu.vlen);
 
-	// compute v=ddiagKi-sum(R.*B, 1)'
-	VectorXd v=ddiagKi-R.cwiseProduct(eigen_B).colwise().sum().adjoint();
+	/*
+	//true posterior mean with equivalent FITC prior
+	//time complexity of the following operations is O(n)
+	Map<VectorXd> eigen_al(m_al.vector, m_al.vlen);
```

Comment:
As mentioned before, we have two posteriors.
* the true posterior p(f_n|y) \propto p(y|f_n)*p(f_n), where p(f_n) is the FITC equivalent FITC prior
* FITC approximated posterior  
p(f_n|y) = \int_{f_m} { p(f_n|f_m)*p(f_m|y) df_{m} }